### PR TITLE
GG-1887 Moving new DI code to separate package and reverting old code back to 3.1.0

### PR DESCRIPTION
**Pull request by WilliamJohnathonLea**

### CODE REVIEW

File: src/main/scala/uk/gov/hmrc/play/config/inject/RunMode.scala [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+ */
+
+package uk.gov.hmrc.play.config.inject
+
+import com.google.inject.Inject
+import play.api.{Application, Configuration, Environment, Mode}
+
+import scala.annotation.tailrec
+
+trait RunMode {
```
Comment (VlachJosef) [inline src/main/scala/uk/gov/hmrc/play/config/inject/RunMode.scala]: This is duplicating all logic from `uk.gov.hmrc.play.config.RunMode`. It should rather extends it.

```
trait RunMode extends uk.gov.hmrc.play.config.RunMode {
```

**Comment by jmaciasluque** [comment]
👍

File: src/main/scala/uk/gov/hmrc/play/config/inject/ServicesConfig.scala [hunk 2]
```diff
@@ -0,0 +1,94 @@
...
+ */
+
+package uk.gov.hmrc.play.config.inject
+
+import com.google.inject.Inject
+import play.api.{Configuration, Environment}
+
+import scala.concurrent.duration.Duration
+
+trait ServicesConfig extends RunMode {
```
Comment (frankoid) [inline src/main/scala/uk/gov/hmrc/play/config/inject/ServicesConfig.scala]: It seems odd that the ServicesConfig logic (setting names, overrides/cascades etc) is duplicated in an inject and a non-inject version of the trait, and it means there's a chance the two versions could diverge in future.

Can they share the logic somehow? Perhaps they can both extend the same trait, or perhaps they can even _be_ the same trait. As long as the mechanism for obtaining the `Configuration` and `Environment` isn't baked in I don't see why the same trait can't be used in both DI and non-DI situations.

File: src/main/scala/uk/gov/hmrc/play/config/RunMode.scala [hunk 3]
```diff
@@ -16,18 +16,14 @@
...
 import scala.annotation.tailrec
 
 trait RunMode {
 
-
-  protected def app: Application
-
-  protected def mode: Mode.Mode = app.mode
-  protected def runModeConfiguration: Configuration = app.configuration
+  protected def mode: Mode.Mode = Play.current.mode
```
Comment (applethepeanut) [inline src/main/scala/uk/gov/hmrc/play/config/RunMode.scala]: could we please drop the needless types?
```
protected def mode = Play.current.mode
```

File: src/main/scala/uk/gov/hmrc/play/config/RunMode.scala [hunk 4]
```diff
@@ -16,18 +16,14 @@
...
+  protected def runModeConfiguration: Configuration = Play.current.configuration
```
Comment (applethepeanut) [inline src/main/scala/uk/gov/hmrc/play/config/RunMode.scala]: remove type

File: src/main/scala/uk/gov/hmrc/play/config/AppName.scala [hunk 5]
```diff
@@ -16,20 +16,12 @@
...
-import com.google.inject.Inject
-import play.api.{Application, Configuration, Play}
+import play.api.{Configuration, Play}
 
 trait AppName {
-
-  protected def app: Application
-
-  protected def appNameConfiguration: Configuration = app.configuration
+  protected def appNameConfiguration: Configuration = Play.current.configuration
```
Comment (applethepeanut) [inline src/main/scala/uk/gov/hmrc/play/config/AppName.scala]: remove type

File: src/main/scala/uk/gov/hmrc/play/config/ServicesConfig.scala [hunk 6]
```diff
@@ -66,11 +67,22 @@ trait ServicesConfig extends RunMode {
...
+  def getConfDuration(confKey: String, defDur: => Duration): Duration =
+    runModeConfiguration.getString(s"$rootServices.$confKey")
+      .orElse(runModeConfiguration.getString(s"$services.$confKey"))
+      .orElse(runModeConfiguration.getString(s"$playServices.$confKey")) match {
+        case Some(s) => Duration.create(s)
+        case None => defDur
+      }
 
-  def getString(key: String) = runModeConfiguration.getString(key).getOrElse(throw new RuntimeException(s"Could not find config key '$key'"))
+  def getInt(key: String): Int = runModeConfiguration.getInt(key).getOrElse(configNotFoundError(key))
```
Comment (applethepeanut) [inline src/main/scala/uk/gov/hmrc/play/config/ServicesConfig.scala]: do we need  the return type when calling the self descriptive `getInt(...)`

File: src/main/scala/uk/gov/hmrc/play/config/ServicesConfig.scala [hunk 7]
```diff
@@ -66,11 +67,22 @@ trait ServicesConfig extends RunMode {
...
 
-  def getBoolean(key: String) = runModeConfiguration.getBoolean(key).getOrElse(throw new RuntimeException(s"Could not find config key '$key'"))
-}
+  def getString(key: String): String = runModeConfiguration.getString(key).getOrElse(configNotFoundError(key))
```
Comment (applethepeanut) [inline src/main/scala/uk/gov/hmrc/play/config/ServicesConfig.scala]: same

File: src/main/scala/uk/gov/hmrc/play/config/ServicesConfig.scala [hunk 8]
```diff
@@ -66,11 +67,22 @@ trait ServicesConfig extends RunMode {
...
+
+  def getBoolean(key: String): Boolean = runModeConfiguration.getBoolean(key).getOrElse(configNotFoundError(key))
```
Comment (applethepeanut) [inline src/main/scala/uk/gov/hmrc/play/config/ServicesConfig.scala]: same

File: src/main/scala/uk/gov/hmrc/play/config/ServicesConfig.scala [hunk 9]
```diff
@@ -66,11 +67,22 @@ trait ServicesConfig extends RunMode {
...
 
-class DefaultServicesConfig @Inject()(val app: Application) extends ServicesConfig
+  def getDuration(key: String): Duration = runModeConfiguration.getString(key).map(Duration.create).getOrElse(configNotFoundError(key))
```
Comment (applethepeanut) [inline src/main/scala/uk/gov/hmrc/play/config/ServicesConfig.scala]: same

File: src/main/scala/uk/gov/hmrc/play/config/inject/RunMode.scala [hunk 10]
```diff
@@ -0,0 +1,35 @@
...
+import com.google.inject.Inject
+import play.api.{Application, Configuration, Environment, Mode}
+
+import scala.annotation.tailrec
+
+trait RunMode extends uk.gov.hmrc.play.config.RunMode {
+
+  protected def environment: Environment
+
+  protected override def mode: Mode.Mode = environment.mode
```
Comment (applethepeanut) [inline src/main/scala/uk/gov/hmrc/play/config/inject/RunMode.scala]: drop type

File: src/main/scala/uk/gov/hmrc/play/config/inject/RunMode.scala [hunk 11]
```diff
@@ -0,0 +1,35 @@
...
+
+}
+
+class DefaultRunMode @Inject()(
+  override val runModeConfiguration: Configuration,
+  val environment: Environment
+) extends RunMode
```
Comment (applethepeanut) [inline src/main/scala/uk/gov/hmrc/play/config/inject/RunMode.scala:35]: git likes a blank new line

File: src/test/scala/uk/gov/hmrc/play/config/ServicesConfigSpec.scala [hunk 12]
```diff
@@ -32,14 +34,17 @@ class ServicesConfigSpec extends WordSpecLike with Matchers with BeforeAndAfterA
...
     "anotherString" -> "hello other test",
-    "anotherBool" -> "false"
+    "anotherBool" -> "false",
+    "anotherDur" -> "60seconds"
   )
 
   lazy val fakeApplication = FakeApplication(additionalConfiguration = config)
 
-  override def beforeAll() {
+  override def beforeAll(): Unit = {
```
Comment (applethepeanut) [inline src/test/scala/uk/gov/hmrc/play/config/ServicesConfigSpec.scala]: pointless return type when the previous version clearly describes this

**Pull request merged**